73-offline-attestation.md

```diff
@@ -0,0 +1,369 @@
+persistent datastore) all the information needed to perform attestation
+ standlone and offline (i.e., quotes and MB/IMA logs).
+- Add functionality on the `verifier` to **optionally** record (in a 
+tamper-resistant transparency log) a hash/signature of the contents 
+of the information needed to perform attestation standalone and offline.
+- Add functionality, on both `verifer` and `registrar` to **optionally**
+request a timestamp from an external Time Stamp Authority whenever storing
+data on the tamper-resistant transparency log.
+- Add a new CLI which will interface with the aforementioned persistent 
+stores, and will call the main, unmodified `verifier` code in order to do
+ post-facto attestation.  
+
+### Non-Goals
+
+<!--
+What is out of scope for this enhancement?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- The interaction between the time-series like persistent datastore and
+  tamper-resistant transparency log will be done by keylime user/operator.
+Inside the core Keylime, a "plugin" architecture will be adopted (similar
+to the "policies" for `Measured Boot`) and the implementation details of the
+code which will interact with such stores are outside of the scope of this
+ enhancement. Nevertheless, some relevant examples of interaction with some
+backends (e.g., Redis, Rekor, "Flat file") will be provided "as is".
+
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+- The `registrar` will be modified to, upon initial `agent` registration -
+  which includes the execution of `tpm2_makecredential` - record the EK, AIK
+into a JSON file, sign it (using the private key generated as part of the
+certificates for mTLS interaction with both `tenant` and `registrar`) and then
+make a record of it on a tamper-resistant transparency log (e.g., Rekor). In
+addition to that, it will store the JSON file, the signature, and the public
+key on the time-series like persistent datastore. This should allow an external
+component/user to check, provided that there is trust on the `registrar`, that
+a particular AIK is indeed tied to a particular EK. The reason for having this
+data stored into a time-series is due to the fact that AIKs are regenereated
+every time an `agent` is restarted on Keylime.
+- The `verifier` will be modified to take the `json_response` (python
+  dictionary) from the `agent` - which will include both TPM quotes and logs (MB
+and IMA) - plus `agent` data (python dictionary which includes all columns stored)
+from the SQL database, internal to Keylime), combine it into a single (python 
+dictionary record) and store it on a time-series datastore.
```

Comment:
The "internal state" of Keylime is needed only in case an external third-party auditor wants to use the `verifier` code **unmodified** to perform offline attestation. Again, there is a real use case for this with our Production customers, but the fact we are storing more, not less data should not prevent other parties to write offline attestation code from scratch, simply ignoring any "Keylime-specific state"